### DISCUSSION TITLE
What do you use openEMS for?

### DISCUSSION THREAD

**Discussion by toammann** [thliebig/openEMS-Project#94] [General]
Hey,
I was wondering: For what kind of projects do you use openEMS for?

Because openEMS is free and open source, I can imagnine some people to use it for hobby projecs. That is acutally my use case. I think the expertise shown in this forum is above hobby use. Do you use openEMS for commercial oder research applications?

Maybe some of you are also interested in this topic. Here is my data:
- What do I do for a living:
I am RF design engineer working in the Aerospace and Test and Measurement industry. I have about 2 years of experience in using commercial EM-Solvers
- What do I use openEMS for?
Hobby projects and also to learn more about FDTD
- Where do I live?
Europe, Germany

Regards,
Tobias

**Comment by JTSvejda** [comment]
Hello,

I'm working at the University of Duisburg-Essen, Germany (https://www.uni-due.de/ate/) and we utilize openEMS primarily within our educational courses:
* Guided antenna design for MRI applications (such as shortened dipoles), or for communications (patch, patch array, dipole, or similar simple antennas)
* In a remote lab, where students have to design a chipless RFID tag. They need to pick a corresponding paper from a pool of 7 papers and reconstruct the structure within openEMS and show that it works for a coded response of the tag via a radar cross section simulation.

Here and there we use openEMS for research but due to the increase of complexity of the research problems, commercial solvers are used mostly. 

Best regards,
Taro

**Reply by 0xCoto** [reply]
Interesting! I'd be curious to learn what kind of challenges you've faced that openEMS falls short of when it comes to research, compared to commercial tools? Is it the solver/engine itself, or the overall usability of the interface, modeling, iterative design process etc.?

**Reply by JTSvejda** [reply]
Generally the installation seems to worked out for every student so far, but they struggle a bit with the scripting interface. From my point of view the biggest problem is to understand how FDTD works in order to set the correct boundary conditions. Furthermore, the manual creation of the mesh is somewhat difficult and especially manifests itself in either clearly too fine meshes, too coarse meshes whose lines do not fit the object structure or with large jumps in the distances.
In both cases, however, the learning effect is high. The students have the opportunity to ask me questions and this works very well via video conference and exchanged script file. We can then address the various problems brought in by the students and I think that is also a very good experience for the use of commercial tools, which solve such matters automatically, although this does not always have to be successful.

**Comment by VolkerMuehlhaus** [comment]
Here's yet another (new) user from Germany :-)

What do I do for a living:
    I started as an RF engineer in 1993, when interoperability of RF design tools was very limited, and created several RF EDA tool interfaces and workflows for my own use. Some of them became real products and my small company switched from design to EDA sales & support, initally focused on RF EDA layout interfaces and now for 25 years specialized in EM simulation using MoM and FDTD.

What do I use openEMS for?
   My EM consulting work uses commercial EM solvers, but then a new project came up: SiGe chip foundry IHP recently started an open source (openPDK) project and asked my to help them finding the best suitable OSS EM solution for that project. After some evaluation, and also due to the lack of suitable MoM solvers, I selected FDTD with openEMS and started creating interfaces to import GDSII layout (used in IC design). Initial benchmarks against measured structures (up to 110 GHz) look good for accuracy, although simulation time is still an issue.

What I really like about openEMS is the scripting capability (some might say: scripting requirement) that gives me a lot of control when building models from layout in a semi-automatic way. The commercial EM tools that I use have fully automated meshing (great) but they don't give me the same level of control in scripting a model. For my current task of prototyping an automated workflow for OSS-based RFIC EM simulation, the openEMS Python solution is great - and fun!

Best regards
Volker

**Reply by leochand101** [reply]
Hi Voker thanks for the feedback , do you think openEMS could be accurate enough to reach lets say 90-95% commercial accuracy on a submicron process <45nm assuming we can define the stack somewhat accurately (also since inductors are basically only dependent mostly on the metal stack ))?? I have an example inductor model training exercise for a library of inductors. The fallback is to use a commercial tool, but I would like to try

**Reply by VolkerMuehlhaus** [reply]
@leochand101 Hi Leo, this is difficult to answer. 

I had shown the feasibility of openEMS for RFIC EM in the examples below, but it really takes much more time compared to commercial EM: 
https://github.com/IHP-GmbH/IHP-Open-PDK/tree/main/ihp-sg13g2/libs.tech/openems

Maybe the biggest issue is the lack of an efficient lossy thick metal model in openEMS, so that we need to mesh into skin effect for thick metal conductors. This means small mesh cells -> small time step -> longer simulation time. 

Sub-micron mesh size at frequencies down to DC, that's what makes FDTD slow for many RFIC modelling tasks. It scales nicely for electrically large models, but for inductors it is clearly slower than FEM or planar MoM.

One difficulty that openEMS shares with all other 3D EM: the ports must be defined with care, because the internal length "inside" a lumped port becomes part of results. In planar MoM like ADS Momentum, it is much easier to get valid results with the port reference placed at the backside of the substrate, because those ports can have a gap between signal and reference. 

But to answer your question: you can get really accurate results if you are willing to go for the small mesh cells and resulting long simulation time.

BR. Volker

**Reply by leochand101** [reply]
If runtime is the only impediment we can just wait longer, or throw more resources at it, as pointed out some of these tools tend to be prohibitive in some cases. I will try to follow your example and give it a try, I am an FDTD noob so if I get it to work I guess most others can too.

**Reply by VolkerMuehlhaus** [reply]
Yeah, maybe give it a try with my inductor model linked above to get an idea about simulation time, and see if the manual definition of mesh lines works for you.

Throwing more resources at the problem is somewhat difficult. openEMS is limited by memory bandwidth and doesn't utilize high end CPU as efficiently as others, according to my experience. I am using 16-core AMD Ryzen 9 5950X simulation server and 8-core Ryzen 7 5800H notebook, which give about the same speed in openEMS. openEMS support multiple cores, but here the memory bandwith is the bottleneck with the openEMS implementation.

But give it a try, my model example and related documentation already worked for others to get their RFIC stuff simulated in openEMS.

**Comment by Etawn** [comment]
Hello,

I have been looking at openEMS for a few weeks but have never dared to ask questions. @VolkerMuehlhaus  and @JTSvejda 's answers to this post were very interesting to me, so I wanted to ask them now. **Please, if my questions are out of scope for this thread or the forum, tell me and I will delete my message.** 

I am a former RF designer for a research center in France. I am mostly doing antennas from a few MHz to 80 GHz using CST or HFSS. I am thinking about starting my own company for antenna design consulting (focus on IoT) and creating products (if I succeed…). I was looking at HFSS/CST startup's prices but for Ansys for example, this license is forbidden for consulting... I don't know about CST. 
For this reason, I was looking at openEMS since it is free and OK with commerical use. Thousand of euros/year is not possible for a licence in a one person company. I am very curious for a comparison with commercial solvers with the goal in mind of designing IoT PCB antennas, antenna arrays or manpack type antennas. 

@JTSvejda , for you, the only problems with openEMS are the scripting, the boundaries and the meshing compared to CST or Ansys ? Other than that, commercial solvers results are close to openEMS ?
- Isn't always putting PML ABC a solution for the boundaries ? I don't know when MUR boundaries should be placed.
- For the last problem, how do you know when it is correctly mesh (especially if there are curves on the PCB for example) ? Can you iterate with an increasing mesh-cells and see if the S11 parameter change >0.1dB for example ? 
- If you have experience with openEMS, when do you know you should change for a commercial solver instead ? When is openEMS not appropriate ?

@VolkerMuehlhaus , you said that you use an EM commercial solver for your company. Is it possible to know which one, please ?

Again if my questions are out of scope for this thread or the forum, tell me and I will delete my message. I did not find answers on the others posts.

Sincerly,

**Reply by VolkerMuehlhaus** [reply]
> @VolkerMuehlhaus , you said that you use an EM commercial solver for your company. Is it possible to know which one, please ?

ADS Momentum and Sonnet (both MoM) and Empire XPU (FDTD). For 3D stuff like co-design of PCB antennas with environment, I use Empire XPU.

**Reply by Etawn** [reply]
Thank you @VolkerMuehlhaus for your very fast answer ! I did not know Empire but it seems to be a good one from my colleagues' opinions. I will definitely give it a look.

**Comment by LubomirJagos42** [comment]
Hi,
first when found openEMS I was surprised that it's capable do microwave simulation right and quite easy but it was missing GUI and there was no equivalent in open-source to HFSS and CST so I tried and made GUI for it to made frontend:
- https://github.com/LubomirJagos42/FreeCAD-OpenEMS-Export
- https://github.com/LubomirJagos42/FreeCAD-OpenEMS-Export-Plugin-Documentation
- https://www.youtube.com/watch?v=OR-79dCMdyo

This should fill the gap between scripting and simulation creation and made it easier for everybody to set simulation as learning openEMS since I suppose that people who want to do simulation don't used to be programmers and programmers don't used to be microwave simulation engineers.

I think that meshing is quite hard part of openEMS or any FDTD software and is not easily to be automatized, I put into GUI way how I used to do meshing, it's just graphical frontend for manual grid definition.

I did it and it helped few people and students to finished they work since then they can focused more on design and not spend months to learn how to do initial scripts just to start, also if somebody is making real complex design scripting could be really tedious and life is too short to spend it inspecting stacktrace :D
